sonic-chassisd/scripts/chassisd

```diff
@@ -167,6 +168,59 @@ class ModuleConfigUpdater(logger.Logger):
             self.log_info("Changing module {} to admin {} state".format(key, 'DOWN' if admin_state == MODULE_ADMIN_DOWN else 'UP'))
             try_get(self.chassis.get_module(module_index).set_admin_state, admin_state, default=False)
 
+#
+# SmartSwitch Module Config Updater ========================================================
+#
+
+
+class SmartSwitchModuleConfigUpdater(logger.Logger):
+
+    def __init__(self, log_identifier, chassis):
+        """
+        Constructor for SmartSwitchModuleConfigUpdater
+        :param chassis: Object representing a platform chassis
+        """
+        super(SmartSwitchModuleConfigUpdater, self).__init__(log_identifier)
+
+        self.chassis = chassis
+        # The threading.Lock is required for concurrent setting of admin_state
+        # on multiple DPUs for faster bring-up and shutdown.
+        self.lock = threading.Lock()
+
+    def deinit(self):
+        """
+        Destructor of SmartSwitchModuleConfigUpdater
+        :return:
+        """
+
+    def module_config_update(self, key, admin_state):
+        if not key.startswith(ModuleBase.MODULE_TYPE_DPU):
+            self.log_error("Incorrect module-name {}. Should start with {}".format(key,
+                                                        ModuleBase.MODULE_TYPE_DPU))
+            return
+
+        module_index = try_get(self.chassis.get_module_index, key, default=INVALID_MODULE_INDEX)
+
+        # Continue if the index is invalid
+        if module_index < 0:
+            self.log_error("Unable to get module-index for key {} to set admin-state {}". format(key, admin_state))
+            return
+
+        if (admin_state == MODULE_ADMIN_DOWN) or (admin_state == MODULE_ADMIN_UP):
+            self.log_info("Changing module {} to admin {} state".format(key, 'DOWN' if admin_state == MODULE_ADMIN_DOWN else 'UP'))
+            # Acquire the lock before submitting the callback function
+            with self.lock:
+                # Submit the callback function as a separate thread. It is the responsibility
+                # of the module to handle multiple b2b requests to the same module
+                t = threading.Thread(target=self.submit_callback, args=(module_index, admin_state))
```

Comment:
Duplicating the comment since it was marked as resolved:
<img width="771" alt="image" src="https://github.com/user-attachments/assets/17f00ecd-246b-4059-949e-e19e7cfc3f65">

Reply:
The lock is valid until the start of the thread. It is safe to have that as multiple threads may access common resources in chassis and modules causing data corruption. Once the thread is created it has its own stack. In fact it is good to have a lock inside the submit_callback.  It is hard to find what common resources are accessed inside each vendor API. So, it is better to have a lock even in submit_callback.